hsigmoid/READ_EN.md

```diff
@@ -0,0 +1,174 @@
+---
+
+# Hsigmoid accelerated training of word vector
+## Background Introduction
+In natural language processing field, usually, using one-hot vector to present word as traditional method, like word dict ['I', 'you', 'like'], [1,0,0],[0,1,0],[0,0,1] three vectors can be used to present 'I','you' and 'like'. Thought this usage is concise, it's easy to produce dimension explosion; And any vectors of two word is orthogonal, so information in vector is limited. To avoid or alleviate weakness of one-hot presentation, now usually we take word vector instead of one-hot presentation, word vector namely word embedding, that is using low-dimension dense vector to replace high-dimension sparse one-hot vector. There are many methos to training word vector, one of them is neural network model, including CBOW,Skip-gram etc. Esentially, these are classification model, when vocabulary, class is large, traditional softmax is much time-consuming. PaddlePaddle applys Hsigmoid Layer, NCE Layer, to accelerate training.Here we mainly introduce how to use Hsigmoid Layer to accelerate training, issues related to word vector please go to PaddlePaddle Book [Chapter word vector](https://github.com/PaddlePaddle/book/tree/develop/04.word2vec)
+
+
+
+##Hsigmoid Layer
+Hsigmoid Layer refered to paper\[[1](#Reference)\], Hsigmoid is Hierarchical-sigmoid, theory of it is to reduce computing complexity by contructing classification binary tree, every leaf node of binary tree is a class,every non-leaf node is a binary classifier. Supposing we get four classes 0,1,2,3, softmax will calculate four classes score seperately, then normalizate it to propability. When classes are large, it's time-consuming to calculate each class propability, Hsigmoid Layer will construct a balanced binary tree based on quantity of classes, as following:
+
+<p align="center">
+<img src="images/binary_tree.png" width="220" hspace='10'/> <img src="images/path_to_1.png" width="220" hspace='10'/> <br/>
+Figure 1. （a）is balanced binary tree，（b）is path from root node to class 1
+</p>
+each non-leaf node is a two-class classifier(sigmoid) in binary tree, if class is 0, then continue judging classes with left child node, if not with right child node, until reach at leaf node. In this mode, each class is a path, like path code if 0, 1 from root to class 1. In training stage, we calculate loss of classifier respectively  by path in order of real class, then summate all loss to final loss. In predicting stage, model will output propability of each non-leaf classifier, we can get path code by propability, then iterate path code to get final predicting class. Computation cost of traditional softmax is N(N is size of vocabulary dict), Hsigmoid can reduce complexity to log(N), detailed theory refers paper \[[1](#Reference)\]。
+
```

Comment:
Hsigmoid Accelerated Word Vector Training

Background 

In the field of natural language processing, traditional practices usually use one-hot vectors to represent words. For example, if the dictionary contains 'I', 'you', and 'like', we can use vectors [1, 0, 0], [0, 1, 0] and [0, 0, 1] to represent 'I', 'you' and 'like' respectively. This representation is relatively simple, but when the vocabulary is large, dimensional explosion can easily occur. Besides, since the vectors of any two words are orthogonal to each other, they contain limited information. In order to avoid or reduce such shortcomings, we use word vectors instead of the one-hot expression. Word vectors are word embedding. That is, a low-dimensional, dense, real vector is used to replace the high-dimensional, sparse, one-hot vector. There are many ways to train word vectors. Neural network models are one of them, including CBOW, Skip-gram, etc. These models are essentially classification models. When the word list and the number of categories are large, softmax will be very time consuming. PaddlePaddle provides Hsigmoid Layer and NCE Layer to accelerate the training process of the model. This article mainly introduces how to use Hsigmoid Layer to speed up training. Please refer to the word vector chapter in PaddlePaddle Book for related content.

Hsigmoid Layer

Hsigmoid Layer is cited from this paper[1]. Hsigmoid refers to Hierarchical-sigmoid. The principle is to reduce the computational complexity by constructing a classification binary tree. Each leaf node in the binary tree represents a category, and each non-leaf node represents a two-category classifier. For example, suppose we have a total of four categories: 0, 1, 2, and 3. Softmax will calculate the scores of the four categories and normalize the results to get a probability. When the number of categories is large, it is very time consuming to calculate the probability of each category. Hsigmoid Layer will construct a balanced binary tree according to the number of categories as follows:

  
Figure 1. (a) Balanced binary tree; (b) Path from root node to category 1

Each non-leaf node in the binary tree is a two-class classifier (sigmoid). If the category is 0, choose the left child node and continue the classification. If not, choose the right child until reaching a leaf node. In this way, each category corresponds to a path. For example, the path from root to category 1 is encoded as "0, 1". In the training stage, we calculate the loss of the corresponding classifier according to the path of each real category, and then synthesize all the losses to obtain the final loss. During the prediction phase, the model outputs the probability of each non-leaf node classifier. We can obtain the path's code based on the probability, and then traverse the path's code to obtain the predicted class. The computational complexity of softmax is N (N is the dictionary size), and Hsigmoid can reduce the complexity to log(N). For detailed theoretical details, refer to this paper [1].